File: bulk_send.js
const XLSX = require("xlsx");

// 1️⃣ Read Excel file
const workbook = XLSX.readFile("iei_processed_members.xlsx");
const sheetName = workbook.SheetNames[0];
const worksheet = workbook.Sheets[sheetName];

// Convert to JSON
let members = XLSX.utils.sheet_to_json(worksheet);

// Normalize fields
members = members.map((m) => {
  return {
    ...m,
    membership_id: String(m.membership_id).trim(),
    phoneno_clean: m.phoneno_clean ? String(m.phoneno_clean) : null,
    phoneno: m.phoneno ? String(m.phoneno) : null,
  };
});

console.log(`Total members found: ${members.length}`);

// 2️⃣ Helper: simulate sending SMS for ONE member
async function sendPasswordForMember(member) {
  const membershipId = member.membership_id;
  const phone =
    member.phoneno_clean && member.phoneno_clean !== "null"
      ? member.phoneno_clean
      : member.phoneno;

  if (!phone) {
    console.log(`[SKIP] No phone for member ${membershipId}, skipping...`);
    return;
  }

  const last4 = String(phone).slice(-4);

  console.log(
    `[SIMULATED] Sending password SMS for ${membershipId} to ${phone} (ending with ${last4})`
  );
}

// 3️⃣ Helper: sleep
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// 4️⃣ Batch sending logic
async function sendInBatches() {
  const BATCH_SIZE = 5;
  const DELAY_MS = 3000;

  let batchNumber = 0;

  for (let i = 0; i < members.length; i += BATCH_SIZE) {
    batchNumber++;
    const batch = members.slice(i, i + BATCH_SIZE);

    console.log(
      `\n=== Sending batch ${batchNumber} (members ${i + 1} to ${
        i + batch.length
      }) ===`
    );

    for (const member of batch) {
      await sendPasswordForMember(member);
    }

    if (i + BATCH_SIZE < members.length) {
      console.log(`Waiting ${DELAY_MS / 1000} seconds before next batch...\n`);
      await sleep(DELAY_MS);
    }
  }

  console.log("\n✅ Done sending messages to all members (simulated).");
}

// 5️⃣ Run
sendInBatches().catch(console.error);

File: bulk_send.py
import pandas as pd
import time
# import requests  # uncomment later when you actually call an API


# 1️⃣ This function will send the message for ONE member
def send_password_for_member(membership_id: str, phone: str):
    """
    Here we will contact the IEI system / API to trigger the SMS.
    For now, we are just simulating by printing.
    """

    # 🔹 TODO: In the future, call the real IEI API here.
    # Example (just an idea, not real):
    # response = requests.post(
    #     "https://iei-portal/send-password",
    #     json={"membership_id": membership_id, "phone": phone}
    # )
    # print(response.status_code, response.json())

    print(f"[SIMULATED] Sending password SMS for {membership_id} to {phone}")


# 2️⃣ Read the processed Excel file
file_path = "iei_processed_members.xlsx"  # must be in the same folder
df = pd.read_excel(file_path)

# Just to be safe, convert phone to string
df["phoneno_clean"] = df["phoneno_clean"].astype(str)

print(f"Total members found: {len(df)}")


# 3️⃣ Decide how many people per batch
BATCH_SIZE = 5        # send to 5 people at a time
DELAY_SECONDS = 3     # wait 3 seconds between batches (you can change this)


# 4️⃣ Loop over the DataFrame in steps of BATCH_SIZE
batch_number = 0

for start_index in range(0, len(df), BATCH_SIZE):
    batch_number += 1
    end_index = start_index + BATCH_SIZE
    batch_df = df.iloc[start_index:end_index]

    print(f"\n=== Sending batch {batch_number} (rows {start_index} to {end_index - 1}) ===")

    # 5️⃣ For each member in this batch, send the SMS
    for _, row in batch_df.iterrows():
        membership_id = row["membership_id"]
        phone = row["phoneno_clean"]

        send_password_for_member(membership_id, phone)

    # 6️⃣ After finishing one batch, wait a few seconds before the next one
    # (But don't wait after the last batch)
    if end_index < len(df):
        print(f"Waiting {DELAY_SECONDS} seconds before next batch...\n")
        time.sleep(DELAY_SECONDS)

print("\n✅ Done sending messages to all members (simulated).")
